### ISSUE TITLE
Performance regression in CI from 6.6 -> 6.7 (and 6.8/7.0)

### ISSUE THREAD

**Issue by jrr** [cypress-io/cypress#15779]
### Current behavior

Cypress got slower in v6.7.

### Desired behavior

That it's about as fast as 6.6. (or faster! 😉 )

### Details

My team is using Cypress with Gitlab CI, and when upgrading from 6.6 to 6.7 our test jobs slowed significantly. (Cypress suites taking 2-3x the time)

I can't share the code, so I produced a new project that demonstrates a performance regression, hopefully for the same reasons:

https://github.com/jrr/cypress-storybook-example

The project contains two Cypress suites - the example tests that came out of the box, and a separate suite that visits [storybook](https://storybook.js.org/) pages (containing miscellaneous Material-UI React components).

Locally (2018 Macbook Pro) the performance difference is negligible. In CI, though, there's noticeable slowdown:

| Suite            | Environment | 6.6                    | 6.7                    | 6.8                    | 7.0                    | 7.2                        |
| ---------------- | ----------- | ---------------------- | ---------------------- | ---------------------- | ---------------------- | -------------------------- |
| Visit Storybook  | Local       | 0:27<br/>0:26          | 0:27<br/>0:27          | 0:31<br/>0:28          | 0:29<br/>0:30          | 0:31<br/>0:30              |
| Visit Storybook  | GitHub      | 0:55<br/>0:54<br/>1:07 | 1:44<br/>1:28<br/>2:05 | 2:04<br/>2:11<br/>1:53 | 1:39<br/>1:45<br/>1:43 | 1:10 <br/> 1:03 <br/> 1:03 |
| Visit Storybook  | GitLab      | 2:54<br/>2:53<br/>2:54 | 5:53<br/>5:15<br/>6:01 | 5:23<br/>5:27<br/>4:57 | 5:09<br/>4:45<br/>4:48 | 4:25 <br/> 3:26 <br/> 4:33 |
| Cypress Examples | Local       | 1:14<br/>1:16          | 1:20<br/>1:21          | 1:21<br/>1:21          | 1:34<br/>1:28          | 1:22 <br/> 1:23            |
| Cypress Examples | GitHub      | 1:31<br/>1:25<br/>1:27 | 1:48<br/>1:40<br/>1:49 | 2:03<br/>2:09<br/>1:42 | 1:45<br/>1:54<br/>1:46 | 1:27 <br/> 1:20 <br/> 1:22 |
| Cypress Examples | GitLab      | 2:03<br/>1:53<br/>1:48 | 3:32<br/>3:32<br/>3:21 | 3:21<br/>3:30<br/>3:04 | 3:28<br/>3:07<br/>3:01 | 2:03 <br/> 1:55 <br/> 2:01 |

(Using Chrome `--headless`. I tried a few runs in case of outliers.)

Both GitHub Actions and GitLab CI are connected to the project repo and logs are publicly visible:

[![CI](https://github.com/jrr/cypress-storybook-example/actions/workflows/build.yml/badge.svg)](https://github.com/jrr/cypress-storybook-example/actions/workflows/build.yml) [![pipeline status](https://gitlab.com/john.ruble/cypress-storybook-example/badges/main/pipeline.svg)](https://gitlab.com/john.ruble/cypress-storybook-example/-/pipelines)

Both jobs are running tests in the `cypress/browsers:node14.16.0-chrome89-ff86` Docker environment.

Gitlab's "[shared runners](https://docs.gitlab.com/ee/user/gitlab_com/index.html#linux-shared-runners)" are rather meager - 3.75GB of RAM and 1 vCPU. GitHub's are slightly [beefier](https://docs.github.com/en/actions/using-github-hosted-runners/about-github-hosted-runners#supported-runners-and-hardware-resources) with 7GB and 2 CPUs.

Here's what's reported by `cypress info`:

```
// GitHub
Cypress Version: 6.6.0
System Platform: linux (Debian - 10.8)
System Memory: 7.29 GB free 1.16 GB
```

```
// Gitlab
Cypress Version: 6.6.0
System Platform: linux (Debian - 10.8)
System Memory: 3.87 GB free 573 MB
```

I'm not sure where the bottleneck is (e.g. CPU, Memory, Disk). Are you aware of performance regression in the last couple of versions? Does Cypress maintain performance tests to catch major deltas from release to release? (it looks like cypress-real-world-app tests may have gotten a bit slower according to [its dashboard](https://dashboard.cypress.io/projects/7s5okt/analytics/run-duration?branches=%5B%7B%22label%22%3A%22develop%22%2C%22suggested%22%3Afalse%2C%22value%22%3A%22develop%22%7D%5D&browsers=%5B%5D&chartRangeMostCommonErrors=%5B%5D&chartRangeSlowestTests=%5B%5D&chartRangeTopFailures=%5B%5D&committers=%5B%5D&cypressVersions=%5B%5D&flaky=%5B%5D&operatingSystems=%5B%5D&runGroups=%5B%5D&specFiles=%5B%5D&status=%5B%7B%22label%22%3A%22Passed%22%2C%22value%22%3A%22PASSED%22%7D%2C%7B%22label%22%3A%22Failed%22%2C%22value%22%3A%22FAILED%22%7D%5D&tags=%5B%5D&timeInterval=WEEK&timeRange=%7B%22startDate%22%3A%222021-01-02%22%2C%22endDate%22%3A%222021-04-02%22%7D&viewBy=TEST_CASE))

Is there anything I can try to further track it down and/or improve it?

## Side Note 1: Retained Prereleases

I started trying to bisect commits between 6.6 and 6.7 to track down what change introduced the performance regression (using the [install pre-release version](https://docs.cypress.io/guides/getting-started/installing-cypress#Install-pre-release-version) guide), but I didn't get far before I hit the end of the prerelease packages.

Could you retain them for longer? Like, say, 3 or 4 releases worth? The prerelease packages are great for trying out the latest bleeding edge, but they could _also_ be great for bisecting history if they were available longer.

## Side Note 2: Cypress Stress Test

If you take the storybook test in this repo, and crank it up from 50 visits to ~300, it starts out pretty zippy:

https://user-images.githubusercontent.com/164652/113465256-a4291a00-9400-11eb-9472-c12df185a037.mp4

But by the end it gets _real slow_:

https://user-images.githubusercontent.com/164652/113465259-abe8be80-9400-11eb-90e0-a09c7dc4af1d.mp4

Do you know what's going on here? I thought it might be useful as a stress/performance test if you don't already have something like this.

---------

_edit: added numbers for Cypress 7.0_
_edit: added numbers for Cypress 7.2_

**Comment by lukeapage** [comment]
We are sensitive to performance changes - our full suite is 8 hours long which we run on 15 2cpu machines in parallel. And we didn’t notice any regression upto 6.8. However 7.0 is 7 times slower than 6.8.0 for us. Probably it is something specific for us 🤷 will be interesting if you identify what went slower for you. I agree that having a few months of commit builds would be awesome. In the past I’ve had to build each commit, pack it, check in and push or else write code that patches a live build in order to try out things.

**Comment by kristianm** [comment]
We also noticed a huge increase in elapsed time in our CI after upgrade to 7.0.0.
I ran some CPU benchmarks against the latest versions. When testing with electron as browser, there's a large difference between 6.8.0 and 7.0.0. When testing with headless chrome, there's a difference starting from 6.7.1, but also a large increase in 7.0.0.

My [benchmark script](https://gist.github.com/kristianm/d2cd6cbf64e5dcab8853046b91b91214) was run against the [cypress kitchensink example repo](https://github.com/cypress-io/cypress-example-kitchensink).
[See my results as comment in the same gist](https://gist.github.com/kristianm/d2cd6cbf64e5dcab8853046b91b91214#gistcomment-3695568).

If CPU usage has been correctly accounted in my tests, then there's a huge increase in CPU usage in 7.0.0 release, especially in sys space. Since the benchmarks were executed on my local laptop with multiple cores, the difference in real elapsed time is not as high as the increase in cpu usage. But in resource-limited environments such as CI containers, the real elapsed time will be affected even more.

**Comment by vishal423** [comment]
This is really annoying as the CI `e2e` tests now takes much more time than before (observed ~3 fold increase). Reference CI [before upgrade](https://github.com/jhipster/generator-jhipster-svelte/runs/2265408519?check_suite_focus=true#step:20:541) and [after upgrade](https://github.com/jhipster/generator-jhipster-svelte/runs/2283775980?check_suite_focus=true#step:20:542)

**Comment by csvan** [comment]
Is there any way to profile exactly what is taking more time?

**Comment by jrr** [comment]
> Is there any way to profile exactly what is taking more time?

@csvan I haven't tried digging in to mine yet, but a couple Gleb blog posts may help:

https://www.cypress.io/blog/2020/05/22/where-does-the-test-spend-its-time/

https://glebbahmutov.com/blog/cypress-timings/

**Comment by kristianm** [comment]
The CPU increase is noticeable in all of the [kitchensink example specs](https://github.com/cypress-io/cypress-example-kitchensink/tree/master/cypress/integration/examples), so it seems more like a core issue than any specific area (network, actions etc).

**Comment by lassesteffen** [comment]
We also noticed a heavy increase in run time in our test suite  (13mins to 28 mins =  2x slower) when upgrading from 6.8.0 to 7.0.0. Very interested in a workaround / fix for this.

**Comment by GC-Mark** [comment]
Can also confirm our tests are taking way longer in `v7.0.0` than the were in `v6.8.0`

Also, because the tests are taking longer, we get random failing tests due to timeout issues.

# Our last few CI runs
<img width="931" alt="Screenshot 2021-04-07 at 10 44 13" src="https://user-images.githubusercontent.com/1477806/113846617-9243d000-978e-11eb-997b-77844d1f8b0b.png">

# v6.8.0
<img width="752" alt="Screenshot 2021-04-07 at 10 47 45" src="https://user-images.githubusercontent.com/1477806/113847015-f5356700-978e-11eb-9559-089ddd77ae01.png">

# v7.0.0
(including a random failing test due to a timeout)
<img width="735" alt="Screenshot 2021-04-07 at 10 47 28" src="https://user-images.githubusercontent.com/1477806/113847002-f23a7680-978e-11eb-8e00-12ca325ba105.png">

**Comment by flotwig** [comment]
@jrr thank you for the comprehensive issue report. Looks like something got messed up between 6.6.0 and 6.7.0.

With regards to your side notes:

> Could you retain them for longer? Like, say, 3 or 4 releases worth? The prerelease packages are great for trying out the latest bleeding edge, but they could also be great for bisecting history if they were available longer.

We retain builds for 31 days, which is typically two release cycles for Cypress.

> But by the end it gets real slow: Do you know what's going on here?

Ok, so, we have seen issues like this before. There is probably some performance regression related to a UI change in the Command Log. As the Command Log grows in size, it takes longer and longer to render, so any minor slowdown in an individual component has a polynomial effect on the total render time.

If I look at the commits between 6.6.0 and 6.7.0: https://github.com/cypress-io/cypress/compare/v6.6.0...v6.7.0 I don't really see anything too obvious that would have caused this, so the best course of action is probably to continue bisecting, building, and testing locally...

**Comment by GC-Mark** [comment]
@flotwig Should we create a separate issue for the performance regression between `6.8.0` and `7.0.0`

**Comment by agg23** [comment]
I can corroborate. Data produced using @kristianm's nice production Cypress bisection script (https://gist.github.com/kristianm/d2cd6cbf64e5dcab8853046b91b91214) running on a 2.4GHz i9 MacBook Pro 16in:

| Cypress Version | Browser  | Browser Version | Node    | Real  | User  | System |
|-----------------|----------|-----------------|---------|-------|-------|--------|
| 6.5.0           | Electron | 11.2.3          | 12.18.3 | 76.64 | 89.85 | 15.80  |
| 6.5.0           | Chrome   | 89.0.4389.114   | 12.18.3 | 79.55 | 31.76 | 6.72   |
| 6.6.0           | Electron | 11.2.3          | 12.18.3 | 78.13 | 88.77 | 15.74  |
| 6.6.0           | Chrome   | 89.0.4389.114   | 12.18.3 | 79.12 | 31.51 | 6.62   |
| 6.7.1           | Electron | 11.3.0          | 12.18.3 | 79.15 | 96.05 | 17.42  |
| 6.7.1           | Chrome   | 89.0.4389.114   | 12.18.3 | 81.45 | 62.91 | 15.15  |
| 6.8.0           | Electron | 11.3.0          | 12.18.3 | 76.68 | 92.77 | 16.84  |
| 6.8.0           | Chrome   | 89.0.4389.114   | 12.18.3 | 80.70 | 63.21 | 14.48  |
| 7.0.0           | Electron | 12.0.2          | 14.16.0 | 94.96 | 199.52| 65.26  |
| 7.0.0           | Chrome   | 89.0.4389.114   | 14.16.0 | 96.30 | 77.03 | 33.96  |

Cypress has not internally noticed this issue, in CI or otherwise, so I wanted to verify that there wasn't something odd with our binary releases. These are run off of the v6.8.0 and v7.0.0 tagged commits:

| Cypress Version | Browser  | Browser Version | Node    | Real  | User  | System |
|-----------------|----------|-----------------|---------|-------|-------|--------|
| 6.8.0           | Electron | 11.3.0          | 12.19.0 | 94.94 | 118.12| 20.05  |
| 6.8.0           | Chrome   | 89.0.4389.114   | 12.19.0 | 103.64| 83.52 | 20.46  |
| 7.0.0           | Electron | 12.0.2          | 14.16.0 | 102.53| 207.92| 71.49  |
| 7.0.0           | Chrome   | 89.0.4389.114   | 14.16.0 | 108.19| 78.92 | 38.99  |

Given the large differences we see in Electron run times vs Chrome, I'm willing to posit that either Electron is getting slower, or we are improperly handling these newer versions of Electron. However, that obviously is not all of the picture...

**Comment by lukeapage** [comment]
@flotwig 
> We retain builds for 31 days, which is typically two release cycles for Cypress.

I am trying to pinpoint the commit where 7.0 starts going slow - so within the last release. and builds appear to be missing older than 14 days, which means I cannot bisect through the last releases history :(

**Comment by flotwig** [comment]
@lukeapage yeah, the only way to do that currently is to `yarn` each commit and `yarn cypress:run` to use the unbuilt of cypress

**Comment by jrr** [comment]
> yeah, the only way to do that currently is to `yarn` each commit and `yarn cypress:run` to use the unbuilt version of cypress

@flotwig Are you describing how to build and use old versions of Cypress? Is this documented anywhere?

I'd consider trying to continue bisecting history without the aid of prerelease packages if it's not too terribly difficult, but I'm not sure where to start. Can you generally build a single tarball of an old version of Cypress, then `yarn install` it in a project?

**Comment by estefafdez** [comment]
Same here, we notice a lot of performance issues moving from 6.9.1 to 7.0 and 7.0.1. 
Any update on how this problem will be fixed? for now, we roll back to 6.9.1.

**Comment by flotwig** [comment]
> Can you generally build a single tarball of an old version of Cypress, then `yarn install` it in a project?

@jrr You *can* do the full build process, but actually you don't even need to do this. 

It's as simple as:

```
# in cypress repo root
yarn # install all dependencies and build

# this is analogous to `cypress open` via the CLI
yarn cypress:open --project ../path/to/your/project
```

Check https://github.com/cypress-io/cypress/blob/develop/CONTRIBUTING.md#common-top-level-tasks for more information

**Comment by jrr** [comment]
@flotwig if my issue were locally reproducible I'd certainly try that! I'm interested in building tarballs for the same reason that I'm interested in longer-retained prerelease packages: so that I can reference one from my project, commit, push, and let CI do its thing.

Will one of those top-level tasks in CONTRIBUTING.md build an installable package? (or if not, is it in another node script like `binary-build` or `build-prod`?)

**Comment by mysticdevx** [comment]
As a contribution to the issue, i have same speed problem when i upgraded from 5.6.0. to 7.1.0. the overall duration is almost tripled. Then just to be sure whether its about my framework, i have run the versions 5.6.0, 6.7.1, and 7.1.0 on the given cypress-kitchen-sink example. below is the summary of the test runs. (p.s. no retries occurred in any test)

<img width="925" alt="runs" src="https://user-images.githubusercontent.com/49609497/115008410-5f46be00-9eab-11eb-9517-eeac28723e03.png">

**Comment by donocode** [comment]
Another data point, I upgraded from 6.3 to 7.1 and saw that test execution time went from ~200 seconds to ~600-700. We run tests periodically in aws fargate.

![snap 210423203008 _8AFTHZ_](https://user-images.githubusercontent.com/6978877/115926660-12795f00-a438-11eb-8606-6f8ad10f373c.png)

**Comment by jrr** [comment]
## Testing Cypress 7.2

I added numbers for 7.2 to the table above, and while much of the recent loss has been regained, it's not _quite_ back to 6.6 performance.

Focusing on the "Visit Storybook" scenario, where Cypress and a webpack dev server are competing for resources, here are the increases in averaged times as a percentage:

Environment | 6.6 | 6.7 | 6.8 | 7.0 | 7.2
 -- | -- | -- | -- | -- | --
Local | (baseline) | 102% | 111% | 111% | 115%
GitHub | (baseline) | 180% | 209% | 174% | 111%
GitLab | (baseline) | 198% | 182% | 169% | 143%

Most importantly, Cypress is now fast enough that I can upgrade my main project without immediately getting a bunch of timeouts and failures. Here are some times from there:

|                 | 6.6                    | 7.2             |
| --------------- | ---------------------- | --------------- |
| storybook       | 5:19<br/>5:38<br/>6:21 | 6:30<br/>6:31   |
| integration 1/4 | 9:56<br/>8:48<br/>8:47 | 12:01<br/>11:38 |
| integration 2/4 | 6:43<br/>6:28<br/>7:04 | 8:26<br/>8:39   |
| integration 3/4 | 6:25<br/>8:01<br/>5:31 | 7:17<br/>7:29   |
| integration 4/4 | 5:09<br/>5:22<br/>5:17 | 7:16<br/>7:33   |

@donocode @mysticdevx how does 7.2 do for you?

**Comment by estefafdez** [comment]
Same here, at least 25 min slower with 7.2.0 than with 6.9.1.

**Comment by redReno** [comment]
We're having the same issue since 6.7.0. 7.2.0 still has the same usse. Tests running in Chrome seem to be the most affected.

**Comment by jennifer-shehane** [comment]
Our team released changes to fix a performance issue introduced in 6.7 onward in this PR: https://github.com/cypress-io/cypress/pull/15985 We believed this should address this issue.

In 7.0, we also increased the default screen size when running Cypress in a headless browser. I believe this increase in size may be accounting for some reduced performance between 6.6 and 7.2.0.  

For those tracking the timings, can you try reducing the screen size to the default size prior to 7.0 (1280x720) to see if the performance issues are related to the increase in screen size? [Instructions here.](https://docs.cypress.io/api/plugins/browser-launch-api#Set-screen-size-when-running-headless)

**Comment by lukeapage** [comment]
We see the performance regressions fixed in 7.2.0 on a 8 hour test run. But we also override the screen size so that change wouldn’t have affected us.

**Comment by timfaase** [comment]
I'm seeing a similar performance drop between 6.6.0 and 7.2.0. From around 33 minutes to 40 minutes. Screen size did not matter.

**Comment by jennifer-shehane** [comment]
Closing as this specific performance regression has been addressed. 

If you are still seeing a decrease in performance, we believe it's from another cause not outlined in this issue. We'd love to track it down if you can provide a reproducible example in this issue: https://github.com/cypress-io/cypress/issues/16305

**Comment by jrr** [comment]
For posterity, I re-ran the repro project with 6.6, and 7.2 a few times at two different resolutions:

| Suite            | Environment | 6.6                    | 7.2 (1280x720)           | 7.2 (3840x2160)          |
| ---------------- | ----------- | ---------------------- | ---------------------- | ---------------------- |
| Visit Storybook  | GitHub      | 0:42<br/>0:57<br/>0:48 | 0:53<br/>1:00<br/>0:51 | 1:14<br/>0:47<br/>1:03 |
| Cypress Examples | GitHub      | 1:12<br/>1:23<br/>1:15 | 1:15<br/>1:21<br/>1:13 | 1:29<br/>1:15<br/>1:26 |
| Visit Storybook  | GitLab      | 3:47<br/>2:52<br/>2:54 | 3:15<br/>3:22<br/>3:09 | 3:06<br/>3:30<br/>3:13 |
| Cypress Examples | GitLab      | 1:54<br/>1:42<br/>1:48 | 1:59<br/>2:15<br/>1:44 | 2:07<br/>2:11<br/>1:59 |

I still think 7.2 is a bit slower than 6.6 (moreso in my private app than in this repro), but it's close enough that it's not a problem anymore. Thanks for your work!
